test/integration/ddl_sharding_key_test.lua

```diff
@@ -0,0 +1,181 @@
+end
+
+local pgroup = t.group('ddl_sharding_key', {
+    {engine = 'memtx'},
+    {engine = 'vinyl'},
+})
+
+pgroup.before_all(function(g)
+    g.cluster = helpers.Cluster:new({
+        datadir = fio.tempdir(),
+        server_command = helpers.entrypoint('srv_ddl'),
+        use_vshard = true,
+        replicasets = helpers.get_test_replicasets(),
+        env = {
+            ['ENGINE'] = g.params.engine,
+        },
+    })
+    g.cluster:start()
+    local result, err = g.cluster.main_server.net_box:eval([[
+        local ddl = require('ddl')
+
+        local ok, err = ddl.get_schema()
+        return ok, err
+    ]])
+    t.assert_equals(type(result), 'table')
+    t.assert_equals(err, nil)
+end)
+
+pgroup.after_all(function(g) helpers.stop_cluster(g.cluster) end)
+
+pgroup.before_each(function(g)
+    helpers.truncate_space_on_cluster(g.cluster, 'customers_name_key')
+    helpers.truncate_space_on_cluster(g.cluster, 'customers_name_key_uniq_index')
+    helpers.truncate_space_on_cluster(g.cluster, 'customers_name_key_non_uniq_index')
+    helpers.truncate_space_on_cluster(g.cluster, 'customers_secondary_idx_name_key')
+end)
+
+pgroup.test_select = function(g)
+    local tuple = {2, box.NULL, 'Ivan', 20}
+
+    -- insert tuple
+    local result, err = g.cluster.main_server.net_box:call('crud.insert', {
+        'customers_name_key', tuple
+    })
+
+    t.assert_equals(err, nil)
+    t.assert_not_equals(result, nil)
+    t.assert_equals(#result.rows, 1)
+
+    local conditions = {{'==', 'id', 2}}
+
+    local result, err = g.cluster.main_server.net_box:call('crud.select', {
+        'customers_name_key', conditions,
+    })
+
+    t.assert_equals(err, nil)
+    t.assert_not_equals(result, nil)
+    t.assert_equals(#result.rows, 1)
+end
+
+-- Right now CRUD's plan for select doesn't support sharding key and it leads
+-- to map reduce (select on all replicasets). To avoid map-reduce one need to
+-- add a separate index by field name, used in select's condition. We plan to
+-- fix this in scope of https://github.com/tarantool/crud/issues/213
+pgroup.test_select_wont_lead_map_reduce = function(g)
+    local space_name = 'customers_name_key_uniq_index'
+    local customers = helpers.insert_objects(g, space_name, {
+        {id = 1, name = 'Viktor Pelevin', age = 58},
+        {id = 2, name = 'Isaac Asimov', age = 72},
+        {id = 3, name = 'Aleksandr Solzhenitsyn', age = 89},
+        {id = 4, name = 'James Joyce', age = 59},
+        {id = 5, name = 'Oscar Wilde', age = 46},
+        {id = 6, name = 'Ivan Bunin', age = 83},
+        {id = 7, name = 'Ivan Turgenev', age = 64},
+        {id = 8, name = 'Alexander Ostrovsky', age = 63},
+        {id = 9, name = 'Anton Chekhov', age = 44},
+    })
+    t.assert_equals(#customers, 9)
+
+    -- Disable vshard's rebalancer and account current statistics of SELECT
+    -- calls on storages before calling CRUD select. Rebalancer may screw up
+    -- statistics of SELECT calls, so we will disable it.
```

Comment:
AFAIU (with @Gerold103's help), the rebalancer will not move buckets in our case. Reasons:

* [3k buckets](https://github.com/tarantool/crud/blob/5e8c91e39bd608804e0ab85989a537dfd63ec7bf/test/entrypoint/srv_select.lua#L201) are created on boostrap (see [vshard.router.bootstrap()](https://github.com/tarantool/vshard/blob/27ef8035732b3585990c838f29651cb2bb8621a8/vshard/router/init.lua#L1330-L1345) and assigned to replicasets uniformly.
* After bootstrap the rebalancer will see ideally balanced cluster (it [looks](https://www.tarantool.io/en/doc/latest/reference/reference_rock/vshard/vshard_architecture/#rebalancer) on bucket counts in each replicaset and don't track space sizes).
* We don't add or remove replicasets, so this ideal will not ruin.

However I don't know for sure, whether it issues periodical SELECT / CALL requests to storages to update knowledge about bucket counts. I'll ask Vlad.

----

I remember that I wrote it here in some comment, but I can't find, so maybe I already asked. If I remember right, you had plans to use crud specific select statistics that was implemented in PR #222. And, if I remember right, there were problems with stability of this test (but I don't know, whether rebalancer disabling helps). So I want to remind about this: whether you want to skip it here and return later, skip it at all or just forgot and want to implement here? Whether the test stable in its current implementation?

Reply:
> However I don't know for sure, whether it issues periodical SELECT / CALL requests to storages to update knowledge about bucket counts. I'll ask Vlad.

Rebalancer polls storages periodically. It calls the [rebalancer_request_state()](https://github.com/tarantool/vshard/blob/27ef8035732b3585990c838f29651cb2bb8621a8/vshard/storage/init.lua#L2436-L2461) function, which performs several SELECTs, which are counted in the `box.stat.SELECT`. So, sure, if we'll lean on the box statistics, we should disable the rebalancer.

Questions above (under the line) are still actual.

Reply:
Discussed it verbally. My implementation of test that relies on `box.stat().SELECT.total` for storage statistics is bad. I'll rewrite it using `helpers.storage_stat.collect()`.